File: hamming-error-correction/src/main/java/com/pivovarit/hamming/vavr/HammingMessageExtractor.java
package com.pivovarit.hamming.vavr;

import io.vavr.Tuple;
import io.vavr.collection.Stream;

class HammingMessageExtractor {
    BinaryString stripHammingMetadata(EncodedString input) {
        String raw = Stream.from(0, 1).take(input.getValue().length())
          .map(i -> Tuple.of(i + 1, Character.toString(input.getValue().charAt(i))))
          .filter(i -> (i._1 & -i._1) != i._1)
          .map(i -> i._2)
          .foldLeft("", String::concat);

        return BinaryString.of(raw);
    }
}

File: hamming-error-correction/src/main/kotlin/com/pivovarit/hamming/domain/decode/HammingMessageExtractor.kt
package com.pivovarit.hamming.domain.decode

import com.pivovarit.hamming.domain.BinaryString
import com.pivovarit.hamming.domain.EncodedString
import com.pivovarit.hamming.domain.isPowerOfTwo

internal class HammingMessageExtractor {
    fun stripHammingMetadata(input: EncodedString): BinaryString {
        return input.value.asSequence()
          .filterIndexed { i, _ -> (i + 1).isPowerOfTwo().not() }
          .joinToString("")
          .let(::BinaryString)
    }
}